File: Tree/Binary Tree Level Order Traversal.cpp
/*
    Company Tags  : Amazon, Microsoft, Bloomberg
    Leetcode Link : https://leetcode.com/problems/binary-tree-level-order-traversal/
*/

// Approach: Breadth-First Search (Level by Level)
// T.C : O(n)
// S.C : O(n)
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (!root) {
            return result;
        }

        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            int levelSize = q.size();
            vector<int> currentLevel;
            currentLevel.reserve(levelSize);

            while (levelSize--) {
                TreeNode* node = q.front();
                q.pop();

                currentLevel.push_back(node->val);

                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
            }

            result.push_back(currentLevel);
        }

        return result;
    }
};

File: BinaryTree/Binary Tree Level Order Traversal.cs
/*
    LeetCode 102: Binary Tree Level Order Traversal
    Approach : Breadth-First Search (Queue)

    Time Complexity  : O(n)
    Space Complexity : O(n)
*/

public class Solution
{
    public IList<IList<int>> LevelOrder(TreeNode root)
    {
        IList<IList<int>> result = new List<IList<int>>();

        if (root == null)
        {
            return result;
        }

        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);

        while (queue.Count > 0)
        {
            int levelSize = queue.Count;
            List<int> level = new List<int>(levelSize);

            for (int i = 0; i < levelSize; i++)
            {
                TreeNode node = queue.Dequeue();
                level.Add(node.val);

                if (node.left != null)
                {
                    queue.Enqueue(node.left);
                }

                if (node.right != null)
                {
                    queue.Enqueue(node.right);
                }
            }

            result.Add(level);
        }

        return result;
    }
}
